lib/telemetry.go

```diff
@@ -338,5 +328,67 @@ func InitTelemetry(cfg TelemetryConfig) (*metrics.InmemSink, error) {
 		metricsConf.EnableHostname = false
 		metrics.NewGlobal(metricsConf, memSink)
 	}
+	return sinks, errors
+}
+
+// InitTelemetry configures go-metrics based on map of telemetry config
+// values as returned by Runtimecfg.Config().
+// InitTelemetry retries configurating the sinks in case error is retriable
+// and retry_failed_connection is set to true.
+func InitTelemetry(cfg TelemetryConfig, logger hclog.Logger) (*metrics.InmemSink, error) {
+	if cfg.Disable {
+		return nil, nil
+	}
+
+	memSink := metrics.NewInmemSink(10*time.Second, time.Minute)
+	metrics.DefaultInmemSignal(memSink)
+	metricsConf := metrics.DefaultConfig(cfg.MetricsPrefix)
+
+	retryWithBackoff := func() {
+		waiter := &retry.Waiter{
+			MaxWait: 5 * time.Minute,
```

Comment:
We talked about potentially making this configurable, at least on the `TelemetryConfig` and not necessarily externally configurable by operators. This could allow us to set it to a short wait in tests and assert that it reached the timeout, maybe? I just wanted to note it here. It may not actually work out without larger changes.

Reply:
Agreed. 

The PR still misses handling the case where the agent process exits and the retry routine is still running. My understanding is that to catch the agent process shutdown signal, we have to extract the [InitTelemetry from the BaseOps method](https://github.com/hashicorp/consul/blob/9acce31237fd9f23363711f4ef25fd1de10fe726/agent/setup.go#L93) and move it after [agent.New](https://github.com/hashicorp/consul/blob/9acce31237fd9f23363711f4ef25fd1de10fe726/command/agent/agent.go#L176). With that, we can pass the agent.ShutdownCh to the retry routine.

Reply:
@eculver , when I revisit your question, I realized that Waiter doesn't have a timeout field https://github.com/hashicorp/consul/blob/5f96b63ae7d663b50b68a03c7b6855194b5119f4/lib/retry/retry.go#L34-L50

In the latest version, a cancel function is added to stop the retry routine on agent exit. Please take a look. Thanks.

Reply:
@eculver , when I revisit your question, I realized that Waiter doesn't have a timeout field https://github.com/hashicorp/consul/blob/5f96b63ae7d663b50b68a03c7b6855194b5119f4/lib/retry/retry.go#L34-L50

In the latest version, a cancel function is added to stop the retry routine on agent exit. Please take a look. Thanks.

Reply:
Thanks for revisiting. I am wondering if rather than morphing it into a "timeout" we would just expose the configuration on the `TelemetryConfig` (eg. `MinFailures`, `MaxWait`, `MinWait`, `Jitter`) so that the callers could configure the behavior if necessary. It is certainly not a blocking concern. I will leave it up to you.